1301-x86-boot-add-MLE-header-and-new-entry-point.patch

```diff
@@ -0,0 +1,107 @@
+From 1af22dc1e7b2200ef8bb178c345d62439758911b Mon Sep 17 00:00:00 2001
+From: Kacper Stojek <kacper.stojek@3mdeb.com>
+Date: Wed, 31 Aug 2022 15:03:51 +0200
+Subject: [PATCH 1301/1323] x86/boot: add MLE header and new entry point
+
+MLE header is used with Intel TXT, together with MB2 headers.
+Entrypoint is different, but it is used just to differentiate
+from other entries by moving a magic number to EAX. Execution
+environment is similar to that of Multiboot 2 and code falls
+through to MB2's entry point.
+
+Signed-off-by: Kacper Stojek <kacper.stojek@3mdeb.com>
+Signed-off-by: Krystian Hebel <krystian.hebel@3mdeb.com>
+---
+ docs/hypervisor-guide/x86/how-xen-boots.rst |  5 +++
+ xen/arch/x86/boot/head.S                    | 45 +++++++++++++++++++++
+ 2 files changed, 50 insertions(+)
+
+diff --git a/docs/hypervisor-guide/x86/how-xen-boots.rst b/docs/hypervisor-guide/x86/how-xen-boots.rst
+index ca77d7c8a3..eb60a1cd80 100644
+--- a/docs/hypervisor-guide/x86/how-xen-boots.rst
++++ b/docs/hypervisor-guide/x86/how-xen-boots.rst
+@@ -55,6 +55,11 @@ If ``CONFIG_PVH_GUEST`` was selected at build time, an Elf note is included
+ which indicates the ability to use the PVH boot protocol, and registers
+ ``__pvh_start`` as the entrypoint, entered in 32bit mode.
+ 
++MLE header is used with Intel TXT, together with MB2 headers. Entrypoint is
++different, but it is used just to differentiate from other entries by moving
++a magic number to EAX. Execution environment is similar to that of Multiboot 2
++and code falls through to ``start``.
++
+ 
+ xen.gz
+ ~~~~~~
+diff --git a/xen/arch/x86/boot/head.S b/xen/arch/x86/boot/head.S
+index 245c859dd7..21ab0edf2c 100644
+--- a/xen/arch/x86/boot/head.S
++++ b/xen/arch/x86/boot/head.S
+@@ -3,6 +3,7 @@
+ #include <public/xen.h>
+ #include <asm/asm_defns.h>
+ #include <asm/fixmap.h>
++#include <asm/intel_txt.h>
+ #include <asm/page.h>
+ #include <asm/processor.h>
+ #include <asm/msr-index.h>
+@@ -113,6 +114,25 @@ multiboot2_header:
+         .size multiboot2_header, . - multiboot2_header
+         .type multiboot2_header, @object
+ 
++        .balign 16
++mle_header:
++        .long   0x9082ac5a  /* UUID0 */
++        .long   0x74a7476f  /* UUID1 */
++        .long   0xa2555c0f  /* UUID2 */
++        .long   0x42b651cb  /* UUID3 */
++        .long   0x00000034  /* MLE header size */
++        .long   0x00020002  /* MLE version 2.2 */
++        .long   (slaunch_stub_entry - start)  /* Linear entry point of MLE (SINIT virt. address) */
++        .long   0x00000000  /* First valid page of MLE */
++        .long   0x00000000  /* Offset within binary of first byte of MLE */
++        .long   0x00000000  /* Offset within binary of last byte + 1 of MLE (set by loader) */
```

Comment:
It can't be set by loader, this is part of measured code so it should be constant, at least for the sake of obtaining hash of Xen binary offline.

Reply:
`(_end - start)`?  It seems to produce `3ba9e0` when `readelf xen` gives:
```
Program Headers:
  Type           Offset   VirtAddr   PhysAddr   FileSiz MemSiz  Flg Align
  LOAD           0x000080 0x00200000 0x00200000 0x2969c0 0x3bb000 RWE 0x40
  NOTE           0x1cefc8 0x001cef48 0x001cef48 0x00024 0x00024 R   0x4
```
`0x3bb000` is just `3ba9e0` rounded up and GRUB seems to subtract min(PhysAddr) from max(PhysAddr + MemSiz) to get the value.

Reply:
Seems about right.

Difference between `3bb000` and `3ba9e0` is that the former is `__2M_rwdata_end`, which [is aligned](https://github.com/xen-project/xen/blob/master/xen/arch/x86/xen.lds.S#L414) to (usually) [page size](https://github.com/xen-project/xen/blob/master/xen/arch/x86/xen.lds.S#L36) (not 2M [because syslinux](https://github.com/xen-project/xen/blob/master/xen/arch/x86/Kconfig#L222-L224)) to separate different page permissions. For the purpose of measuring the code, `_end` will be sufficient.

Reply:
Updated: https://github.com/QubesOS/qubes-vmm-xen/compare/5bf7fa7ea0c8887586a878d393fe9e5fb6d55d71..d7feb432f70b9a258f9b07ccc27d7c21b13337e1#diff-4926d144f2f3be8d097b3998d79d9dcd75adf0d06fbf573b6fc52797143d5315R62

Related PRs:
 * https://github.com/TrenchBoot/grub/pull/18
 * https://github.com/TrenchBoot/xen/pull/9